src/qtics/experiment.py

```diff
@@ -0,0 +1,206 @@
+        if not data_file.endswith(".hdf5"):
+            data_file = f"{name}_{time.strftime('%m_%d_%H_%M_%S')}.hdf5"
+        if data_dir != "":
+            os.makedirs(data_dir, exist_ok=True)
+        self.data_file = os.path.join(data_dir, data_file)
+        self.inst_names = []
+        for attr_name in dir(self):
+            if (
+                not attr_name.startswith("_")
+                and Instrument in type(getattr(self, attr_name)).__mro__
+            ):
+                self.inst_names.append(attr_name)
+        if hasattr(self, "__annotations__"):
+            for attr_name, attr_type in self.__annotations__.items():
+                if Instrument in attr_type.__mro__:
+                    self.inst_names.append(attr_name)
+
+    def __del__(self):
+        """Disconnect all devices and delete."""
+        self.all_instruments("clear_defaults")
+        self.all_instruments("disconnect")
+
+    @abstractmethod
+    def main(self):
+        """Execute main part of the experiment."""
+
+    def run(self):
+        """Run the experiment."""
+        log.info("Starting experiment %s.", self.name)
+        try:
+            self.main()
+        except KeyboardInterrupt:
+            log.warning("Interrupt signal received, exiting")
+            self.all_instruments("reset")
+            raise KeyboardInterrupt
+        except Exception as e:
+            log.error("\n\nException occured: %s", e)
+            self.all_instruments("reset")
+            raise Exception(e)
+        log.info("Experiment run successfully.")
+
+    def add_instrument(self, inst: Instrument):
+        """Add an instrument with the correct name."""
+        name = inst.name
+        if name not in self.inst_names:
+            log.warning(
+                "Instrument name %s not in allowed instruments %s.",
+                name,
+                self.inst_names,
+            )
+        else:
+            setattr(self, name, inst)
+            log.info("Added instrument %s.", name)
+
+    def all_instruments(self, func_name, *args, **kwargs):
+        """Apply function to all instruments."""
+        for name in self.inst_names:
+            if hasattr(self, name):
+                inst = getattr(self, name)
+                func = getattr(inst, func_name)
+                _ = func(*args, **kwargs)
+
+    def append_data_group(
+        self, group_name: str, parent_name="", datasets=None, **attributes
+    ):
+        """Save data appending to hdf5 file."""
+        with h5py.File(self.data_file, "a") as file:
+            if parent_name != "":
+                parent_group = file.require_group(parent_name)
+                group = parent_group.require_group(group_name)
+            else:
+                group = file.require_group(group_name)
+
+            if datasets is not None:
+                for data_name, data in datasets.items():
+                    group.create_dataset(data_name, data=data)
+            if attributes:
```

Comment:
In general, it is recommended to not rely on "bare" truthfulness and falseness of arguments to check if they exist or not.